### DISCUSSION TITLE
Mysql provider filter with _id issue. "Unknown column _id in where clause"

### DISCUSSION THREAD

**Discussion by gmentzik** [Budibase/budibase#8205] [Help]
Hi all,

1) I am using Mysql as datasource and my tables have a PK column named id. Could this be conflicting in any way with the auto generated id column of Budibase? 
2) I have a custom page including an edit form with provider selecting the row to be edit with the Mysql id (url/:id <- MySQL id). 
However if I use filtering with _id (Budi autogenerated id) it say no such column exists!
If I auto generate the edit page _id works! 

So before I start getting crazy what could be wrong here?
And finally is there any advantate using the _id (Budibase id) vs Mysql(Datasourse) Id?

**Comment by melohagan** [comment]
Hey @gmentzik 

The Budibase **_id** will simply map to the PK of external datasources, so in this case **_id** would map to *id*.

I'm surprised that Budibase cannot find **_id** when filtering. Could you provide some screenshots and/or the app export please?

**Reply by gmentzik** [reply]
I did the following test.
I copied the data provider component from auto generated page edit page to my custom edit page and replaced the original.
Kept at bottom of page the old provider and also added a new one with same filter as the one I copied from the autogenerated page.
The provider copied from autogenerated page works fine.
The old and the new generate following error when loading page: Unknown column 'environment._id' in 'where clause'
, where environment is the table name in MySQL data source.
My version is 2.0.18. Can you reproduce the error?

**Reply by gmentzik** [reply]
Besides that error I am facing.
My questions remain: 
1. Is there a possibility of conflict betweek mysql id and autogenerated budibase id?
2. Besides row update where _id is necessary to address the mysql row we want to update, is there any advantage (besides that of a more cryptic id at url bar), filtering by _id instead of id of mysql?

**Reply by gmentzik** [reply]
@melohagan 
I have been able to reproduce the problem with _id many times:
At the attached application: 
1. Created a mysql table smt_testtable and added some rows
2. Autgenerated pages for it
3. At edit page I added a 2nd provider under the autogenerated and filtered by _id


Sorry github doesn let me upload a capture.
But the result is that my 2nd provider generates again  Unknown column 'smt_testtable._id' in 'where clause' error.

**Reply by melohagan** [reply]
And what happens if you filter on **smt_id** instead of **_id** ?

**Reply by gmentzik** [reply]
> And what happens if you filter on **smt_id** instead of **_id** ?

Filtering with the smt_id column works fine provided the link to that page is using this value as url/:id.

**Reply by melohagan** [reply]
I'm not able to do much with the app export because I can't connect to the DB, I should have thought of that before I asked for the app export!

So if you were to add a Data Provider -> Repeater -> Paragraph and print out **_id** and **smt_id** do they look the same, or is there a difference?

**Reply by gmentzik** [reply]
> I'm not able to do much with the app export because I can't connect to the DB, I should have thought of that before I asked for the app export!
> 
> So if you were to add a Data Provider -> Repeater -> Paragraph and print out **_id** and **smt_id** do they look the same, or is there a difference?

![image](https://user-images.githubusercontent.com/65217733/195296108-629c1b2e-0420-4352-88cd-e72c3bb4b460.png)

1st provider is the auto generated. Haven't touched filter. It works both passing smt_id and _id value at url. (Didn't know that)
2nd provider is the user added, set to filter with smt_id.

Regarding mysql db table is nothing special. A simple table with : smt_id (int) pk AI column and a value (varchar(45) column.

**Reply by melohagan** [reply]
Ah okay I see, when using **_id** its applying encoding to the id. Hmm, looks related to: https://github.com/Budibase/budibase/issues/6194

**Reply by gmentzik** [reply]
> Ah okay I see, when using **_id** its applying encoding to the id. Hmm, looks related to: #6194
Here is with filtering on _id 
![image](https://user-images.githubusercontent.com/65217733/195318293-deb533c0-2ea3-49a9-8312-639f2feded1e.png)
I don't think encoding is the issue. Something different there is between auto generated provider and user added provider.

**Reply by gmentzik** [reply]
@melohagan  Any news? I feel unsafe with the solution to create an auto generated page to copy provider so that _id works. What if it breaks down in future?
Also even if I use the auto created provider, if i change or delete the filter with _id and re introduce it generates again same error. It tries to find a column with that name instead of using the magic Budibase id.
